### DISCUSSION TITLE
When a TXD serial port is sending data, how do I change a pin to low level and automatically change the pin to high level after the data is sent

### DISCUSSION THREAD

**Discussion by ShiPC123** [stm32duino/Arduino_Core_STM32#2307] [General]
I used the following code
```C++
void setup() {
Serial1.begin(115200);
pinMode(PA0, OUTPUT);
digitalWrite(PA0, HIGH);
}

void loop() {
digitalWrite(PA0, LOW);
Serial1.println("12345678");
digitalWrite(PA0, HIGH);
}
```

But the `delay()` function must be added if it doesn't seem to work properly
Like this

```C++
void setup() {
Serial1.begin(115200);
pinMode(PA0, OUTPUT);
digitalWrite(PA0, HIGH);
}

void loop() {
digitalWrite(PA0, LOW);
Serial1.println("12345678");
delay(5);
digitalWrite(PA0, HIGH);
}
```

Is there a way to do this without adding a delay function

**Comment by fpistm** [comment]
USe `Serial.flush();` instead of `delay(5)`.

```C++
void setup() {
Serial1.begin(115200);
pinMode(PA0, OUTPUT);
digitalWrite(PA0, HIGH);
}

void loop() {
digitalWrite(PA0, LOW);
Serial1.println("12345678");
Serial1.flush();
digitalWrite(PA0, HIGH);
}
```

For ref: https://www.arduino.cc/reference/tr/language/functions/communication/serial/flush/

**Comment by ShiPC123** [comment]
I don't want to use delay because I want to use a non-blocking way to run the function, for example, I have a baud rate of 9600 and now I want to send 50 frames of data and it takes about 60ms. I don't want my code stuck here, is there any other way?

**Reply by fpistm** [reply]
This is not a delay, this simply ensure the data is fully sent.  When you do a Serial print, this fill the buffer and start send action which is managed by IT, that's all.
Don't understand what is your issue here, if you don't want to wait the data is fully sent then simply do not flush or delay...

**Reply by ShiPC123** [reply]
If I use Serial1.flush(); The program will be stuck here until the data is completely sent. I send a lot of data at one time, so the code will be stuck here for a long time, which will affect the execution of other code

**Reply by fpistm** [reply]
So don't use the `Serial1.flush()` and continue your application, data will be sent in all case.

**Reply by ShiPC123** [reply]
The communication chip I use is RS485. When the data is sent, DE_IO must be pulled down, otherwise the data cannot be sent. Therefore, I need such a function, but I do not want to implement it in a blocking way.

**Reply by ShiPC123** [reply]
[
![RS485switchingdirection02](https://github.com/stm32duino/Arduino_Core_STM32/assets/132763025/2a3026b9-b9f4-40c4-b79c-76f08b83ab06)
](url)

**Reply by fpistm** [reply]
There is no service to warn when all data sent in Arduino API (ie: callback).

**Reply by fpistm** [reply]
Just thinking of `availableForWrite` API, you can try to call it periodically (using HardwareTimer) and if the result equal the buffer size (`SERIAL_TX_BUFFER_SIZE`), it means all data were sent and so you can toggle your pin.
It is just thought, I have not tested.

**Reply by ShiPC123** [reply]
Thank you very much

**Reply by ShiPC123** [reply]
I found the following library file. How should I interpret the transmitDisableIRQ () function inside
https://github.com/PR-DC/PRDC_RS485HD_STM32/blob/main/src/PRDC_RS485HD_STM32.cpp

void PRDC_RS485HD_STM32::transmitDisableIRQ(serial_t *obj) {
it = objs.find(&obj-&gt; index);
PRDC_RS485HD_STM32* oRS485 = it-&gt; second;
if(oRS485-&gt; state == 1) {
oRS485-&gt; state = 0;
if(oRS485-&gt; pin_num) {
digitalWrite(oRS485-&gt; enable_pin,LOW);
if(oRS485-&gt; pin_num &gt;  1) {
digitalWrite(oRS485-&gt; enable_pin2,LOW);
}
}
}
}

**Reply by fpistm** [reply]
@ShiPC123, you can easily understand I can't study all user code. The only thing I can tell is that the library is for this core and it seems manage your use case to manage the pin. Up to you to use it or not or asking support to the library maintainer.
